[magic-value-comparison] Custom allowlist for Magic Numbers (PLR2004)

DaniBodor opened the pull request:
<!--
Thank you for contributing to Ruff/ty! To help us out with reviewing, please consider the following:

- Does this pull request include a summary of the change? (See below.)
- Does this pull request include a descriptive title? (Please prefix with `[ty]` for ty pull
  requests.)
- Does this pull request include references to any relevant issues?
-->

## Summary

This PR allows users to specify a custom set of whitelisted values for magic value comparisons in PLR2004. The setting is named `pylint.allow_magic_values`, which is consistent with a similar setting `pylint.allow_magic_value_types`
The previous hard-coded allow-list was moved to the default for `pylint.allow_magic_values`, giving users full control over what is and what is not a whitelisted value. This setting is added in addition to the `pylint.allow_magic_value_types` and does not replace it. Any value that is whitelisted due to either of these settings is not considered a linting violation.

Note that as far as I could tell, no other setting allows floats or complex numbers as inputs, so I created a few helper methods to deal with these properly (mainly relating to floating point variation).


## Test Plan

I have tested the settings with a number of cases manually in a repository I work on regularly. I have not added any extensive (combinatorial) tests to the test suite.

## Note

I don't know whether you have any policy on using or disclosing the use of code generators, but I definitely relied heavily on this to solve some of the issues I ran into during development. Ultimately, all code was reviewed by me before pushing.

fix: #10009

Commit:
allow magic values

Commit:
move allowed list to defaults

Commit:
clarify documentation of PLR2004

MichaReiser commented on crates/ruff_linter/src/rules/pylint/settings.rs:
```diff
@@ -5,8 +5,132 @@ use serde::{Deserialize, Serialize};
+    Int(i64),
+    Float(f64),
+    Complex { real: f64, imag: f64 },
+    Bytes(Vec<u8>),
```
What's the syntax for specifying a bytes literal value in the settings?

Commit:
only allow strings and numbers as magic value exceptions

Commit:
create AllowedFloatValue struct

DaniBodor committed:
configurations

DaniBodor committed:
update defaults

DaniBodor committed:
snapshots & schema

DaniBodor committed:
use matches_literal

instead of try_from_literal_expr

DaniBodor committed:
allow matching whole number floats against ints

i.e. listing an allowed magic value of the float `1.0` will also accept the int `1` (and vice versa)

DaniBodor committed:
setting to extend default allow list

in addition to defining custom list (which replaces the defaults)

DaniBodor commented on crates/ruff_linter/src/rules/pylint/settings.rs:
```diff
@@ -5,8 +5,132 @@ use serde::{Deserialize, Serialize};
+}
+
+impl AllowedValue {
+    pub fn try_from_literal_expr(literal_expr: LiteralExpressionRef<'_>) -> Option<Self> {
```
addressed in https://github.com/astral-sh/ruff/pull/18961/commits/08dec84944e8272523fa78dbe7311a4032b3d9bf

DaniBodor commented:
> This is still on my list to review!

just a friendly ping because it's been a few weeks now

ntBre reviewed (changes requested):
Thanks, this looks pretty reasonable to me at a high level. I had a few suggestions for simplifying a bit further, and I think we should add some tests. At a minimum we should test the somewhat tricky logic in `AllowedValue::matches` with tests in crates/ruff_linter/src/rules/pylint/mod.rs, but we may also want to include a CLI test or two for parsing from an actual TOML file.

DaniBodor committed:
clippy suppression using expect

Co-authored-by: Brent Westbrook <36778786+ntBre@users.noreply.github.com>

DaniBodor committed:
remove excess documentation

DaniBodor committed:
refactor allowed magic value matching to per-value literal checks

DaniBodor commented on crates/ruff_linter/src/rules/pylint/rules/magic_value_comparison.rs:
```diff
@@ -73,29 +77,31 @@ fn as_literal(expr: &Expr) -> Option<LiteralExpressionRef<'_>> {
-        LiteralExpressionRef::BytesLiteral(_) => true,
+        LiteralExpressionRef::StringLiteral(_)
+        | LiteralExpressionRef::NumberLiteral(_)
+        | LiteralExpressionRef::BytesLiteral(_) => true,
```
Yes, that's definitely more concise. I think the compromise between conciseness and explicitness would be to invert the logic and explicitly encode the cases intended to be ignored:

`!matches!(literal_expr, LiteralExpressionRef::NoneLiteral(_) | LiteralExpressionRef::BooleanLiteral(_) | LiteralExpressionRef::EllipsisLiteral(_))`

DaniBodor committed:
simplify ignoring `None`, `Bool`, and `Ellipsis`

DaniBodor committed:
simplify comparing floats to ints

DaniBodor committed:
extend custom allowlists

DaniBodor committed:
remove obsolete functionality

DaniBodor reviewed (commented):
Thanks for the review!
Sorry for initially commenting on individual points, rather than batching them in one review.

I think tests make a lot of sense, I use a dummy toml file for one of the repos I used to work on as a test. In doing so, I actually discovered that there's a bug with negative values, which I will need to look into.
I'll also look into creating tests, but may need some help if I get stuck.

In the meantime, I think I've addressed all your other comments already. I will wait with re-opening this PR until I have finished some tests and fixing that negative value bug, but in principle am not working anymore your other comments.

DaniBodor committed:
fix negatives bug

DaniBodor committed:
refactor `matches_value`

DaniBodor commented:
@ntBre @MichaReiser @amyreese , just a gentle ping as I haven't had an update for a while now

DaniBodor commented:
@ntBre or @MichaReiser , any chance I can get an update on the cause of the stall and on if and when this PR will be re-reviewed?

ntBre commented on crates/ruff_workspace/src/options.rs:
```diff
@@ -3293,6 +3293,32 @@ pub struct PylintOptions {
+    /// in addition to the default set.
+    ///
+    /// This option is used to extend the default set without replacing it.
+    /// If `allow-magic-values` is set, this option is ignored.
```
Is this resolved? It looks like the comment is still the same, even if the behavior changed.

ntBre commented on crates/ruff_linter/src/rules/pylint/settings.rs:
```diff
@@ -5,8 +5,123 @@ use serde::{Deserialize, Serialize};
+    match number {
+        Number::Int(i) => i.as_i32(),
+        #[expect(clippy::cast_possible_truncation)]
+        Number::Float(f) if f.fract() == 0.0 => Some(*f as i32),
```
I think this will give the wrong result for floats outside the i32 range. See this [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=7fa7319c6b61a0d38054b21e79ce7fba) example. Apologies if this was changed as a result of my previous review.

ntBre commented on crates/ruff_linter/src/rules/pylint/settings.rs:
```diff
@@ -5,8 +5,123 @@ use serde::{Deserialize, Serialize};
+}
+
+impl AllowedValue {
+    pub fn matches_value(
```
I still find this to be a really nice idea, to handle the equality of ints and floats, but I also feel like it adds a layer of unnecessary complexity. How would you feel about omitting this, at least in a first version of the feature, and just comparing ints to ints and floats to floats? It will be slightly more annoying for users if they need to add both `0` and `0.0` to their configs, for example, but at least it's a one-time cost, and it should really simplify the implementation.

ntBre reviewed (changes requested):
Thanks again for working on this and sorry for the delay in review. Things have been quite busy for us in the past couple of months.

I'm leaning toward avoiding the complexity of the numeric comparisons for now. See my inline comment for a little more detail. 

Otherwise, we still need to add tests for this feature. I recommend using the new mdtest feature I just landed. Documented here:

https://github.com/astral-sh/ruff/blob/cc6fa6616ea5dbe93fc84677add62e402ce6f92d/CONTRIBUTING.md?plain=1#L233-L240
and with an example here:

https://github.com/astral-sh/ruff/blob/cc6fa6616ea5dbe93fc84677add62e402ce6f92d/crates/ruff_linter/resources/mdtest/flake8-bandit.md?plain=1#L1-L4

This will be really useful because you'll be able to test the TOML parsing directly in the same file as the tests.

DaniBodor committed:
remove obsolete comment

DaniBodor committed:
Revert "simplify comparing floats to ints"

This reverts commit 89b4f76620091c6495ad079793a9f2b588401dc6.

DaniBodor committed:
Revert "allow matching whole number floats against ints"

This reverts commit 627cbae856ce2155065b7ce4ffa0d7567737e404.

DaniBodor committed:
remove obsolete functions

DaniBodor committed:
run `cargo insta test --force-update-snapshots`

DaniBodor commented:
> I'm leaning toward avoiding the complexity of the numeric comparisons for now. See my inline comment for a little more detail.

Hi @ntBre ,
I finally had some time on my hands once again to look back into this. Given that there's been quite some changes to ruff since my last edits, I decided to rebase the entire branch off of the most recent main ruff branch. I had to do some manual conflict fixes, but am reasonably confident it all went ok.

I agree that cross-checking ints against floats and vice versa probably adds excessive complexity for how useful it is. In fact, I could even imagine cases where a user does want to white-list the one but not the other. So indeed probably best to remove it. I reverted the the code that does the cross-checking in the last few commits.

When pushing the newest changes, I keep running into failed CI checks for which I can't figure out what is wrong. Locally, when I run any of the diagnostics listed in the [dev documentation](https://github.com/astral-sh/ruff/blob/main/crates/ty/CONTRIBUTING.md#development) it passes just fine.
Any chance you can help me diagnose this? FYI, 588afd4 was added in response to [this failed check](https://github.com/astral-sh/ruff/actions/runs/31103964627/job/92624323317) on the previous commit, but seems to have auto-triggered additional reviewers to this PR (once again).

ntBre commented:
Thanks for the update!

The CI failure I'm seeing is for `ruff_dev::bin/ruff_dev generate_json_schema::tests::test_generate_json_schema`, which means you need to run the `cargo dev generate-all` command suggested (way earlier 😅) in the output. The suggestion to run `cargo insta test --force-update-snapshots` is to fix a warning that won't break CI, so we should drop that commit for now.

The contributing docs you linked to are for ty. This is probably the better reference for Ruff, including the mention of `cargo dev generate-all`: https://github.com/astral-sh/ruff/blob/main/CONTRIBUTING.md#example-adding-a-new-configuration-option.

DaniBodor commented:
> Thanks for the update!
> 
> The CI failure I'm seeing is for `ruff_dev::bin/ruff_dev generate_json_schema::tests::test_generate_json_schema`, which means you need to run the `cargo dev generate-all` command suggested (way earlier 😅) in the output. The suggestion to run `cargo insta test --force-update-snapshots` is to fix a warning that won't break CI, so we should drop that commit for now.
> 
> The contributing docs you linked to are for ty. This is probably the better reference for Ruff, including the mention of `cargo dev generate-all`: https://github.com/astral-sh/ruff/blob/main/CONTRIBUTING.md#example-adding-a-new-configuration-option.

Thanks for pointing me to the correct contribution guidelines, not sure how I ended up on the ty guide, it's been the one I was using from the start, which explains why similar things happened a few time 🤦.

I now dumped the `force-update-snapshots` commit and instead ran the json schema update and that was indeed all that was required to make the tests pass.
This PR is now ready for review again

DaniBodor commented:
There was a comment on the original issue regarding the name of these rules that hasn't been addressed in this PR or the discussion yet.

https://github.com/astral-sh/ruff/issues/10009#issuecomment-3006582886 says:

> If the rule's named magic-value-comparison (PLR2004) then an option lint.pylint.allow-magic-values seems appropriate.
> There's already lint.pylint.allow-magic-value-types, so "allow" rather than "allowed" and "values" rather than "value" would keep a similar pattern.
> 
> Although that does bring into question whether lint.pylint.allow-magic-value-types and lint.pylint.allow-dunder-method-names should have been called allowed-* because it's a list of options, not a boolean. As the pattern other rule options seem to follow.

I agree that using `allowed-*` for all these settings is semantically better, but I guess the question is whether it is worth creating a breaking change with the previous existing rule-name.

EDIT: 
Another point that hasn't been addressed (yet) from the issue that this is based off, is whether `2` and `3` should be part of the default allowed list. This was discussed prior to proposing to create an adaptable setting. I think that it would make sense to add `2`, but not `3`. I won't touch the PR now, in case that would make it move down your to do list, but if you agree I'll add it in the next round of changes.
